### PR TITLE
`Communication`: Disable notifications for hidden conversations

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Users receive notifications for conversations that are hidden by a user, which can be annoying for a user since they do not seem to be interested in the conversation.

### Description

If a conversation is hidden, users don't receive notifications anymore, except it is an announcement channel.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Instructor
- Conversations of each type (standard channel, course wide channel, group chat, one to one chat)

1. Login as the student
2. Hide each of the conversations
3. Leave the messages tab with the student
4. Login as the instructor as well
5. Write messages as the instructor, the student should not see any notifications
6. Leave the messages tab with the instructor
7. Write messages as the student, the instructor should see notifications
8. Create a new lecture with a visible date in the future
9. Write messages within the lecture channel as the instructor, the student should not see notifications

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ConversationWebSocketRecipientSummary.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5339/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/ConversationWebSocketRecipientSummary.html) | 85% | ✅ |
| [UserRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5339/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/UserRepository.html) | 91% | ✅ |
| [ConversationMessagingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5339/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/ConversationMessagingService.html) | 87% | ✅ |
| [PostingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5339/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/PostingService.html) | 99% | ✅ |
| [ConversationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5339/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ConversationService.html) | 96% | ✅ |

### Screenshots

No UI changes

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java [hunk 1]
```diff
@@ -115,37 +113,72 @@ public Post createMessage(Long courseId, Post newMessage) {
...
+        conversationService.notifyAllConversationMembersAboutNewMessage(course, conversation, broadcastRecipients);
 
         // creation of message posts should not trigger entity creation alert
         // Websocket notification 3
+        var notificationRecipients = filterNotificationRecipients(author, conversation, webSocketRecipients);
         conversationNotificationService.notifyAboutNewMessage(createdMessage, notificationRecipients, course);
     }
 
+    /**
+     * Filters the given list of recipients for users that should receive a notification about a new message.
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java:150]:
```suggestion
     * Filters the given list of recipients for users that should receive a notification about a new message.
     * <p>
```

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 2]
```diff
@@ -447,18 +447,28 @@ public Set<User> findUsersInDatabase(@RequestBody List<String> userLogins) {
...
-            else if (channel.getExam() != null) {
-                return channel.getExam().isVisibleToStudents();
-            }
-            return true;
-        });
+        return channels.filter(this::isChannelVisibleToStudents);
+    }
+
+    /**
+     * Determines whether the provided channel is visible to students. If the channel is not associated with a
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:454]:
```suggestion
     * Determines whether the provided channel is visible to students.
     * <p>
     * If the channel is not associated with a
```

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 3]
```diff
@@ -447,18 +447,28 @@ public Set<User> findUsersInDatabase(@RequestBody List<String> userLogins) {
...
+     * lecture/exam/exercise, then this method returns true. If it is connected to a lecture/exam/exercise, then the
+     * channel visibility depends on the visible date of the lecture/exam/exercise.
+     *
+     * @param channel the channel under consideration
+     * @return true if the channel is visible to students
+     */
+    public boolean isChannelVisibleToStudents(Channel channel) {
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:461]:
```suggestion
    public boolean isChannelVisibleToStudents(@NonNull Channel channel) {
```

**Review by MaximilianAnzinger — COMMENTED** [review]
Left some nitpicky suggestions. Other than that the code looks good.

**Review by MaximilianAnzinger — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 4]
```diff
@@ -121,18 +116,23 @@ OR lower(user.login) = lower(:#{#searchInput}))
     Set<User> findAllInGroup(@Param("groupName") String groupName);
 
     @Query("""
-            SELECT DISTINCT user
+            SELECT NEW de.tum.in.www1.artemis.domain.ConversationWebSocketRecipientSummary (
+                user,
+                CASE WHEN cp.isHidden = true THEN true ELSE false END,
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:121]: Is the case select needed here?
```suggestion
                cp.isHidden
```

File: src/main/java/de/tum/in/www1/artemis/repository/metis/ConversationParticipantRepository.java [hunk 5]
```diff
@@ -37,6 +37,14 @@ public interface ConversationParticipantRepository extends JpaRepository<Convers
             """)
     Set<ConversationParticipant> findConversationParticipantByConversationId(@Param("conversationId") Long conversationId);
 
+    @EntityGraph(type = LOAD, attributePaths = { "user.groups", "user.authorities" })
+    @Query("""
+            SELECT DISTINCT conversationParticipant
+            FROM ConversationParticipant conversationParticipant
+            WHERE conversationParticipant.conversation.id = :#{#conversationId}
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/metis/ConversationParticipantRepository.java:44]:
```suggestion
            WHERE conversationParticipant.conversation.id = :conversationId
```

**Review by MaximilianAnzinger — APPROVED** [review]
Code still looks good, tested manually on ts2.

**Review by Kroko-fant — APPROVED** [review]
Tested on ts2 in a testing session

**Review by nityanandaz — APPROVED** [review]
Manual test on TS2

A hidden conversation did not show a notification in the bottom right corner

**Review by basak-akan — APPROVED** [review]
Tested during the testing session on TS2, worked as expected.

**Review by Kroko-fant — APPROVED** [review]
tested during the testing session

**Review by Kroko-fant — APPROVED** [review]
reviewed in a testing session on ts2 worked fine 👍

**Review by dearjasmina — APPROVED** [review]
code looks good

**Review by bassner — APPROVED** [review]

**Pull request merged**